modules/nlu-testing/src/tools/visualisation.ts

```diff
@@ -0,0 +1,189 @@
+  Object.entries(groupedIntentsTrain).map(([k, v]: [string, any[]], i) =>
+    traces.push({
+      x: v.map(o => pcaTrain.predict([o.utt_emb]).get(0, 0)),
+      y: v.map(o => pcaTrain.predict([o.utt_emb]).get(0, 1)),
+      z: v.map(o => pcaTrain.predict([o.utt_emb]).get(0, 2)),
+      mode: 'markers',
+      type: 'scatter3d',
+      name: k,
+      text: v.map(o => o.utt),
+      marker: { size: 8, color: i }
+    })
+  )
+
+  if (state.testDatas.length) {
+    const pcaTest = new PCA(state.testDatas.map(o => o.utt_emb))
+    const varianceTest = pcaTest.getExplainedVariance()
+    logger.info(
+      `Top 3 test variance ${varianceTest.slice(0, 3).map(o => _.round(o, 2))} Accounting for ${_.round(
+        _.sum(varianceTest.slice(0, 3)),
+        2
+      )}%`
+    )
+    const groupedIntentsTest = _.groupBy(state.testDatas, 'intent')
+    Object.entries(groupedIntentsTest).map(([k, v]: [string, any[]], i) =>
+      traces.push({
+        x: v.map(o => pcaTest.predict([o.utt_emb]).get(0, 0)),
+        y: v.map(o => pcaTest.predict([o.utt_emb]).get(0, 1)),
+        z: v.map(o => pcaTest.predict([o.utt_emb]).get(0, 2)),
+        mode: 'markers',
+        type: 'scatter3d',
+        name: `${k}_test`,
+        text: v.map(o => o.utt),
+        marker: { size: 8, color: i }
+      })
+    )
+  }
+  return traces
+}
+
+export async function computeKmeansPairwiseIntent(state: BotState) {
+  const grouped_intents = _.groupBy(state.trainDatas, 'intent')
+  const simMat = { matrix: [], labels: [], text: [] }
+  for (const [intent, o] of Object.entries(grouped_intents)) {
+    const rowMat = []
+    const rowText = []
+    for (const [intentTodo, oTodo] of Object.entries(grouped_intents)) {
+      if (intent === intentTodo) {
+        rowMat.push(0)
+        rowText.push('')
+      } else {
+        let bestBadUttsText = []
+        let bestBadUttsNb = 10000
+        for (let i = 0; i < 10; i++) {
+          const kmeans = new clustering.KMEANS()
+          const [cluster1, cluster2] = kmeans.run(
+            o.concat(oTodo).map(o => o.utt_emb),
+            2
+          )
+          const clusterO = _.mean(cluster1) < _.mean(cluster2) ? cluster1 : cluster2
+          const badUttsText = []
+          let badUttsNb = 0
+          o.map((p, i) => {
+            if (!clusterO.includes(i)) {
+              badUttsText.push(p.utt)
+              badUttsNb += 1
+            }
+          })
+          if (badUttsNb < bestBadUttsNb) {
+            bestBadUttsNb = badUttsNb
+            bestBadUttsText = badUttsText
+          }
+        }
+        rowMat.push(bestBadUttsNb)
+        rowText.push(bestBadUttsText.join('<br>'))
+      }
+    }
+    simMat.labels.push(intent)
+    simMat.matrix.push(rowMat)
+    simMat.text.push(rowText)
+  }
+  const plotlyMatrixData = [
+    {
+      x: simMat.labels,
+      y: simMat.labels,
+      z: simMat.matrix,
+      text: simMat.text,
+      type: 'heatmap'
+    }
+  ]
+  return plotlyMatrixData
+}
```

Comment:
50 lines with absolutely no new lines? Usually I suggest grouping const together, add spaces before and after a for statement, and anywhere where it makes it more readable